keras_rs/src/metrics/ranking_metric.py

```diff
@@ -0,0 +1,200 @@
+    according to their relevance or utility for a given query or context.
+    These metrics provide quantitative measures of how well a model achieves
+    this goal, typically by comparing the predicted order of items against the
+    ground truth relevance judgments for each list.
+
+    To define your own ranking metric, subclass this class and implement the
+    `compute_metric` method.
+
+    Args:
+        k: int. The number of top-ranked items to consider (the 'k' in 'top-k').
+            Must be a positive integer.
+        name: Optional name for the loss instance.
+        dtype: The dtype of the metric's computations. Defaults to `None`, which
+            means using `keras.backend.floatx()`. `keras.backend.floatx()` is a
+            `"float32"` unless set to different value
+            (via `keras.backend.set_floatx()`). If a `keras.DTypePolicy` is
+            provided, then the `compute_dtype` will be utilized.
+    """
+
+    def __init__(self, k: Optional[int] = None, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+
+        if k is not None and (not isinstance(k, int) or k < 1):
+            raise ValueError(
+                f"`k` should be a positive integer. Received: `k` = {k}."
+            )
+
+        self.k = k
+
+    @abc.abstractmethod
+    def compute_metric(
+        self,
+        y_true: types.Tensor,
+        y_pred: types.Tensor,
+        mask: types.Tensor,
+        sample_weight: types.Tensor,
+    ) -> types.Tensor:
+        """Abstract method, should be implemented by subclasses."""
+        pass
+
+    def update_state(
+        self,
+        y_true: types.Tensor,
+        y_pred: types.Tensor,
+        sample_weight: Optional[types.Tensor] = None,
+    ) -> None:
+        """
+        Accumulates statistics for the ranking metric.
+
+        Args:
+            y_true: tensor. Ground truth values. Of shape `(list_size)`
+                for unbatched inputs or `(batch_size, list_size)` for batched
+                inputs.
+            y_pred: tensor. The predicted values, of shape `(list_size)` for
+                unbatched inputs or `(batch_size, list_size)` for batched
+                inputs. Should be of the same shape as `y_true`.
+            sample_weight: float/tensor. Can be float value, or tensor of
+                shape `(list_size)` or `(batch_size, list_size)`. Defaults to
+                `None`.
+        """
+        # TODO (abheesht): Should `y_true` be a dict, with `"mask"` as one key
+        # for parity  with pairwise losses?
+
+        # === Convert to tensors, if list ===
+        # TODO (abheesht): Figure out if we need to cast tensors to
+        # `self.dtype`.
+        if isinstance(y_true, list):
+            y_true = ops.convert_to_tensor(y_true)
+        if isinstance(y_pred, list):
+            y_pred = ops.convert_to_tensor(y_pred)
+        # `sample_weight` can be a scalar too.
+        if isinstance(sample_weight, (list, float, int)):
```

Comment:
Remove all 3 `if isinstance(...)` and just call `ops.convert_to_tensor`, it does the if for you.